### PR TITLE
Deserialize arbitrary objects

### LINKED ISSUE #1293: Extending serialization for various types
It would be great to be able to work with arbitrary types in model components. However, the serialization to dict 
constraint becomes a bit of a pain. 

This will require some additional thought on how to work with it. It might be different to what
we have now, but it should be possible to make it work.

The serialization part is easy. Might just require a `to_dict` method. 

```python
class GoodEnough: 
    dims: Dims

    def create_variable(self, name: str) -> TensorVariable:
        ...
        return ...

    def to_dict(self) -> dict[str, Any]: 
        return { ... }
```

However, how will the deserialization work? 

```python
def from_dict(data: dict[str, Any]): 
    ...

from_dict({
    "type": "GoodEnough",
    "data": { ... }
})
```

All of our objects and more should be recovered from the dict.

Would our current objects require to be serialized in a different way? 

For instance, each looks a bit different

```python
# Prior class
Prior("Normal", mu=Prior("Normal")).to_json()
# {'dist': 'Normal', 'kwargs': {'mu': {'dist': 'Normal'}}}

# Media transformation
GeometricAdstock(priors={"alpha": 1}, l_max=12).to_dict()
# {'lookup_name': 'geometric',
#  'prefix': 'adstock',
#  'priors': {'alpha': 1.0},
#  'l_max': 12,
#  'normalize': True,
#  'mode': 'After'}
```
Maybe each is different enough to detect differences but extended ones will
have to follow a certain pattern to detect the type.

### PR DESCRIPTION
**Pull request by wd60622**
## Description

Adding a `deserialize` function which will work on any object registered.
Requires a two step process of checking for the object then deserializing it.

For example, if I would like to deserialize some arbitrary class: 

```python
from pymc_marketing.deserialize import register_deserialization, deserialize

class AbitraryClass:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def to_dict(self):
        """Not required here but just to show what is being deserialized."""
        return {"a": self.a, "b": self.b}


register_deserialization(
    is_type=lambda data: "a" in data and "b" in data,
    deserialize=lambda data: AbitraryClass(**data),
)


# Then using
instance = deserialize({"a": 1, "b": 2})
```

> [!NOTE]
> This two step process comes from the fact that the current serialized objects have different structures from each other

## Related Issue

- [ ] Closes #1293
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [x] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--1296.org.readthedocs.build/en/1296/

### CODE REVIEW

**Comment by wd60622** [comment]
Closes #1300

**Review by juanitorduz — APPROVED** [review]
LGTM! Left some comments on doctstrings and maybe thinking of a better name for "is_type" ? (just to see if there is a better alternative, otherwise is fine)

File: pymc_marketing/deserialize.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+IsType = Callable[[Any], bool]
+Deserialize = Callable[[Any], Any]
+
+
+@dataclass
+class Deserializer:
+    """Object to store a deserialization mapping."""
+
+    is_type: IsType
+    deserialize: Deserialize
```
[the remark this answers is not in the document]
Comment (wd60622) [reply pymc_marketing/deserialize.py]: Added more to the documentation. Mention the 2 step process of deserialization the various data we use

File: pymc_marketing/deserialize.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    else:
+        raise DeserializableError(data)
+
+
+def register_deserialization(is_type: IsType, deserialize: Deserialize) -> None:
+    """Register a deserialization mapping.
+
+    Parameters
+    ----------
+    is_type : Callable[[Any], bool]
```
[the remark this answers is not in the document]
Comment (wd60622) [reply pymc_marketing/deserialize.py:206]: Main entry point will be `deserialize` and `register_deserialization` functions. `register_deserialization` has the `is_type` parameter which has more details and examples. 

I am not attached to the name. `data_checker` is also a bit unclear to me. It sounds like a class instance over function

**Comment by juanitorduz** [comment]
Thanks @wd60622 !

**Pull request merged**